java.lang.VerifyError: Bad type on operand stack in $XControllerTest$Definition$Exec.dispatch; Type 'java/lang/Object' (current frame, stack[0]) is not assignable to integer

### Expected Behavior

This error occurs when a test is run, in this case it is an API functional test for a controller. The expectation is that the test should run and either report success or failure.

### Actual Behaviour

When the test attempts to run, it fails before execution. The test report says "Gradle Test Executor 2. failed to execute tests" (the number of the executor varies).  The full error is similar to the following:

```text
org.gradle.api.internal.tasks.testing.TestSuiteExecutionException: Could not complete execution for Gradle Test Executor 2.
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:63)
	at java.base@11.0.12/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base@11.0.12/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base@11.0.12/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base@11.0.12/java.lang.reflect.Method.invoke(Method.java:566)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:36)
	at org.gradle.internal.dispatch.ReflectionDispatch.dispatch(ReflectionDispatch.java:24)
	at org.gradle.internal.dispatch.ContextClassLoaderDispatch.dispatch(ContextClassLoaderDispatch.java:33)
	at org.gradle.internal.dispatch.ProxyDispatchAdapter$DispatchingInvocationHandler.invoke(ProxyDispatchAdapter.java:94)
	at com.sun.proxy.$Proxy5.stop(Unknown Source)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker$3.run(TestWorker.java:193)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.executeAndMaintainThreadName(TestWorker.java:129)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:100)
	at org.gradle.api.internal.tasks.testing.worker.TestWorker.execute(TestWorker.java:60)
	at org.gradle.process.internal.worker.child.ActionExecutionWorker.execute(ActionExecutionWorker.java:56)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:133)
	at org.gradle.process.internal.worker.child.SystemApplicationClassLoaderWorker.call(SystemApplicationClassLoaderWorker.java:71)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.run(GradleWorkerMain.java:69)
	at app//worker.org.gradle.process.internal.worker.GradleWorkerMain.main(GradleWorkerMain.java:74)
Caused by: org.junit.platform.commons.JUnitException: TestEngine with ID 'junit-jupiter' failed to discover tests
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:111)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discover(EngineDiscoveryOrchestrator.java:85)
	at org.junit.platform.launcher.core.DefaultLauncher.discover(DefaultLauncher.java:92)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:75)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.processAllTestClasses(JUnitPlatformTestClassProcessor.java:99)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor$CollectAllTestClassesExecutor.access$000(JUnitPlatformTestClassProcessor.java:79)
	at org.gradle.api.internal.tasks.testing.junitplatform.JUnitPlatformTestClassProcessor.stop(JUnitPlatformTestClassProcessor.java:75)
	at org.gradle.api.internal.tasks.testing.SuiteTestClassProcessor.stop(SuiteTestClassProcessor.java:61)
	... 18 more
Caused by: org.junit.platform.commons.JUnitException: ClassSelector [className = 'com.keap.forms.api.admin.controller.$AdminControllerTest$Definition$Exec'] resolution failed
	at org.junit.platform.launcher.listeners.discovery.AbortOnFailureLauncherDiscoveryListener.selectorProcessed(AbortOnFailureLauncherDiscoveryListener.java:39)
	at app//org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolveCompletely(EngineDiscoveryRequestResolution.java:103)
	at app//org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.run(EngineDiscoveryRequestResolution.java:83)
	at app//org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolver.resolve(EngineDiscoveryRequestResolver.java:113)
	at app//org.junit.jupiter.engine.discovery.DiscoverySelectorResolver.resolveSelectors(DiscoverySelectorResolver.java:46)
	at app//org.junit.jupiter.engine.JupiterTestEngine.discover(JupiterTestEngine.java:69)
	at org.junit.platform.launcher.core.EngineDiscoveryOrchestrator.discoverEngineRoot(EngineDiscoveryOrchestrator.java:103)
	... 25 more
Caused by: java.lang.VerifyError: Bad type on operand stack
Exception Details:
  Location:
    com/company/proj1/api/admin/controller/$XControllerTest$Definition$Exec.dispatch(ILjava/lang/Object;[Ljava/lang/Object;)Ljava/lang/Object; @329: invokestatic
  Reason:
    Type 'java/lang/Object' (current frame, stack[0]) is not assignable to integer
  Current Frame:
    bci: @329
    flags: { }
    locals: { 'com/company/proj1/api/admin/controller/$XControllerTest$Definition$Exec', integer, 'java/lang/Object', '[Ljava/lang/Object;', '[Z' }
    stack: { 'java/lang/Object' }
  Bytecode:
    0000000: b802 0c3a 041b aa00 0000 0414 0000 0000
    0000010: 0000 0031 0000 00d6 0000 00e5 0000 00f4
    0000020: 0000 0103 0000 0112 0000 0137 0000 014d
    0000030: 0000 015c 0000 016b 0000 017b 0000 018b
    0000040: 0000 019b 0000 01aa 0000 01bc 0000 01cf
    0000050: 0000 01e4 0000 01f9 0000 020d 0000 0221
    0000060: 0000 0233 0000 0246 0000 0258 0000 0267
    0000070: 0000 0276 0000 0285 0000 0294 0000 02a3
    0000080: 0000 02b2 0000 02c1 0000 02d0 0000 02df
    0000090: 0000 02ee 0000 02fd 0000 030c 0000 031b
    00000a0: 0000 032a 0000 0339 0000 0348 0000 0357
    00000b0: 0000 0366 0000 0375 0000 0384 0000 0394
    00000c0: 0000 03a4 0000 03b4 0000 03c4 0000 03d4
    00000d0: 0000 03e4 0000 03f4 0000 0404 2c59 c000
    00000e0: 0db6 0165 0119 0405 0454 b02c 59c0 000d
    00000f0: b601 6701 1904 0604 54b0 2c59 c000 0db6
    0000100: 0169 0119 0407 0454 b02c 59c0 000d b601
    0000110: 6b01 1904 0804 54b0 2c59 c000 0d2d 0332
    0000120: c000 5f2d 0432 c000 5f2d 0532 c000 69b6
    0000130: 016f b601 7201 1904 1006 0454 b02c 2a08
    0000140: b601 76b2 017c b801 82b8 0186 1904 1007
    0000150: 0454 b02c 59c0 000d b601 8919 0410 0804
    0000160: 54b0 2c59 c000 0db6 018c 1904 1009 0454
    0000170: b02c 59c0 0082 b601 8e01 1904 100a 0454
    0000180: b02c 59c0 0082 b601 9001 1904 100b 0454
    0000190: b02c 59c0 0082 b601 9201 1904 100c 0454
    00001a0: b02c 59c0 0082 b601 9519 0410 0d04 54b0
    00001b0: 2c2a 100c b601 762d b801 8219 0410 0e04
    00001c0: 54b0 2c2a 100d b601 762d b801 8201 1904
    00001d0: 100f 0454 b02c 59c0 0082 2d03 32c0 00af
    00001e0: b601 9819 0410 1004 54b0 2c59 c000 822d
    00001f0: 0332 c000 afb6 019a 1904 1011 0454 b02c
    0000200: 2a10 10b6 0176 b201 7cb8 0182 1904 1012
    0000210: 0454 b02c 2a10 11b6 0176 b201 7cb8 0182
    0000220: 1904 1013 0454 b02c 2a10 12b6 0176 2db8
    0000230: 0182 1904 1014 0454 b02c 2a10 13b6 0176
    0000240: 2db8 0182 0119 0410 1504 54b0 2c2a 1014
    0000250: b601 762d b801 8219 0410 1604 54b0 2c59
    0000260: c000 82b6 019d 1904 1017 0454 b02c 59c0
    0000270: 0082 b601 a019 0410 1804 54b0 2c59 c000
    0000280: 82b6 01a3 1904 1019 0454 b02c 59c0 0082
    0000290: b601 a619 0410 1a04 54b0 2c59 c000 82b6
    00002a0: 01a9 1904 101b 0454 b02c 59c0 0082 b601
    00002b0: ac19 0410 1c04 54b0 2c59 c000 82b6 01af
    00002c0: 1904 101d 0454 b02c 59c0 0082 b601 b219
    00002d0: 0410 1e04 54b0 2c59 c000 82b6 01b5 1904
    00002e0: 101f 0454 b02c 59c0 0082 b601 b819 0410
    00002f0: 2004 54b0 2c59 c000 82b6 01bb 1904 1021
    0000300: 0454 b02c 59c0 0082 b601 be19 0410 2204
    0000310: 54b0 2c59 c000 82b6 01c1 1904 1023 0454
    0000320: b02c 59c0 0082 b601 c419 0410 2404 54b0
    0000330: 2c59 c000 82b6 01c7 1904 1025 0454 b02c
    0000340: 59c0 0082 b601 ca19 0410 2604 54b0 2c59
    0000350: c000 82b6 01cd 1904 1027 0454 b02c 59c0
    0000360: 0082 b601 d019 0410 2804 54b0 2c59 c000
    0000370: 82b6 01d3 1904 1029 0454 b02c 59c0 0082
    0000380: b601 d619 0410 2a04 54b0 2c59 c001 46b6
    0000390: 01d8 0119 0410 2b04 54b0 2c59 c001 46b6
    00003a0: 01da 0119 0410 2c04 54b0 2c59 c001 46b6
    00003b0: 01dc 0119 0410 2d04 54b0 2c59 c001 46b6
    00003c0: 01de 0119 0410 2e04 54b0 2c59 c001 46b6
    00003d0: 01e0 0119 0410 2f04 54b0 2c59 c001 46b6
    00003e0: 01e2 0119 0410 3004 54b0 2c59 c001 46b6
    00003f0: 01e4 0119 0410 3104 54b0 2c59 c001 46b6
    0000400: 01e6 0119 0410 3204 54b0 2c59 c001 46b6
    0000410: 01e8 0119 0410 3304 54b0 2a1b b601 ec19
    0000420: 0410 3404 54bf                         
  Stackmap Table:
    append_frame(@220,Object[#526])
    same_frame(@235)
    same_frame(@250)
    same_frame(@265)
    same_frame(@280)
    same_frame(@317)
    same_frame(@339)
    same_frame(@354)
    same_frame(@369)
    same_frame(@385)
    same_frame(@401)
    same_frame(@417)
    same_frame(@432)
    same_frame(@450)
    same_frame(@469)
    same_frame(@490)
    same_frame(@511)
    same_frame(@531)
    same_frame(@551)
    same_frame(@569)
    same_frame(@588)
    same_frame(@606)
    same_frame(@621)
    same_frame(@636)
    same_frame(@651)
    same_frame(@666)
    same_frame(@681)
    same_frame(@696)
    same_frame(@711)
    same_frame(@726)
    same_frame(@741)
    same_frame(@756)
    same_frame(@771)
    same_frame(@786)
    same_frame(@801)
    same_frame(@816)
    same_frame(@831)
    same_frame(@846)
    same_frame(@861)
    same_frame(@876)
    same_frame(@891)
    same_frame(@906)
    same_frame(@922)
    same_frame(@938)
    same_frame(@954)
    same_frame(@970)
    same_frame(@986)
    same_frame(@1002)
    same_frame(@1018)
    same_frame(@1034)
    same_frame(@1050)

	at java.base/java.lang.Class.getDeclaredMethods0(Native Method)
	at java.base/java.lang.Class.privateGetDeclaredMethods(Class.java:3166)
	at java.base/java.lang.Class.privateGetPublicMethods(Class.java:3191)
	at java.base/java.lang.Class.getMethods(Class.java:1904)
	at org.junit.platform.commons.util.ReflectionUtils.getDefaultMethods(ReflectionUtils.java:1520)
	at org.junit.platform.commons.util.ReflectionUtils.getDeclaredMethods(ReflectionUtils.java:1493)
	at org.junit.platform.commons.util.ReflectionUtils.findMethod(ReflectionUtils.java:1343)
	at org.junit.platform.commons.util.ReflectionUtils.isMethodPresent(ReflectionUtils.java:1244)
	at org.junit.jupiter.engine.discovery.predicates.IsTestClassWithTests.hasTestOrTestFactoryOrTestTemplateMethods(IsTestClassWithTests.java:50)
	at org.junit.jupiter.engine.discovery.predicates.IsTestClassWithTests.test(IsTestClassWithTests.java:46)
	at org.junit.jupiter.engine.discovery.ClassSelectorResolver.resolve(ClassSelectorResolver.java:67)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.lambda$resolve$2(EngineDiscoveryRequestResolution.java:135)
	at java.base/java.util.stream.ReferencePipeline$3$1.accept(ReferencePipeline.java:195)
	at java.base/java.util.ArrayList$ArrayListSpliterator.tryAdvance(ArrayList.java:1632)
	at java.base/java.util.stream.ReferencePipeline.forEachWithCancel(ReferencePipeline.java:127)
	at java.base/java.util.stream.AbstractPipeline.copyIntoWithCancel(AbstractPipeline.java:502)
	at java.base/java.util.stream.AbstractPipeline.copyInto(AbstractPipeline.java:488)
	at java.base/java.util.stream.AbstractPipeline.wrapAndCopyInto(AbstractPipeline.java:474)
	at java.base/java.util.stream.FindOps$FindOp.evaluateSequential(FindOps.java:150)
	at java.base/java.util.stream.AbstractPipeline.evaluate(AbstractPipeline.java:234)
	at java.base/java.util.stream.ReferencePipeline.findFirst(ReferencePipeline.java:543)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolve(EngineDiscoveryRequestResolution.java:189)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolve(EngineDiscoveryRequestResolution.java:126)
	at org.junit.platform.engine.support.discovery.EngineDiscoveryRequestResolution.resolveCompletely(EngineDiscoveryRequestResolution.java:92)
	... 30 more
```

### Steps To Reproduce

The tests run to completion when using Micronaut 3.3.4.  After upgrading to either 3.4.4 or 3.5.3, the errors occur every time the tests are run. We have our tests split into unit tests and integration tests. The unit tests run fine in all cases. Only the integration tests fail, and only within a specific subproject/module.  The command that fails is:

```
gradlew :service:api:integrationTest
```

I have spent a great deal of time trying to figure out what the root cause is, and how to reproduce, but so far the only thing that fixes it is rolling back to Micronaut 3.3. If I figure anything else out I will post an update.

Some characteristics of the test in question:

- Junit 5
- Groovy with abstract Java superclass from a library compiled against Micronaut 3.4.4
- `@MicronautTest(transactional = false)`
- `@TestInstance(TestInstance.Lifecycle.PER_CLASS)`

### Environment Information

Tested with Windows 10 and Linux
JDK version: 11

### Example Application

_No response_

### Version

3.4.4 / 3.5.3

--- a/inject-groovy/src/test/groovy/io/micronaut/aop/compile/ExecutableSuperclassSpec.groovy
+++ b/inject-groovy/src/test/groovy/io/micronaut/aop/compile/ExecutableSuperclassSpec.groovy
@@ -1,0 +1,33 @@
+package io.micronaut.aop.compile
+
+import io.micronaut.ast.transform.test.AbstractBeanDefinitionSpec
+import io.micronaut.inject.BeanDefinition
+
+class ExecutableSuperclassSpec extends AbstractBeanDefinitionSpec {
+
+    void "test super class properties are properly recognized"() {
+        when:
+        BeanDefinition beanDefinition = buildBeanDefinition('test.SuperClassFactory', '''
+package test
+
+import io.micronaut.context.annotation.*
+
+@Factory
+class SuperClassFactory extends io.micronaut.aop.compile.SuperClass {
+}
+''')
+
+        then:
+        noExceptionThrown()
+        beanDefinition != null
+
+        when:
+        Object instance = beanDefinition.class.classLoader.loadClass('test.SuperClassFactory').newInstance()
+
+        then:
+        beanDefinition.findMethod("myBool").get().invoke(instance) == false
+        beanDefinition.findMethod("myInt").get().invoke(instance) == 12
+        beanDefinition.findMethod("myDouble").get().invoke(instance) == 15.5
+    }
+
+}

--- a/inject-groovy/src/test/groovy/io/micronaut/aop/compile/SuperClass.java
+++ b/inject-groovy/src/test/groovy/io/micronaut/aop/compile/SuperClass.java
@@ -1,0 +1,20 @@
+package io.micronaut.aop.compile;
+
+import io.micronaut.context.annotation.Executable;
+
+@Executable
+public class SuperClass {
+
+    protected boolean myBool() {
+        return false;
+    }
+
+    protected int myInt() {
+        return 12;
+    }
+
+    public double myDouble() {
+        return 15.5;
+    }
+
+}

--- a/inject/src/main/java/io/micronaut/inject/writer/DispatchWriter.java
+++ b/inject/src/main/java/io/micronaut/inject/writer/DispatchWriter.java
@@ -500,6 +500,7 @@ public class DispatchWriter extends AbstractClassFileWriter implements Opcodes {
             ClassElement returnType = methodElement.isSuspend() ? ClassElement.of(Object.class) : methodElement.getReturnType();
             boolean isInterface = declaringType.getType().isInterface();
             Type returnTypeObject = JavaModelUtils.getTypeReference(returnType);
+            boolean hasArgs = !argumentTypes.isEmpty();
 
             // load this
             writer.loadArg(1);
@@ -507,21 +508,16 @@ public class DispatchWriter extends AbstractClassFileWriter implements Opcodes {
             if (reflectionRequired) {
                 writer.loadThis();
                 writer.push(methodIndex);
-                writer.invokeVirtual(
-                        ExecutableMethodsDefinitionWriter.SUPER_TYPE,
-                        GET_ACCESSIBLE_TARGET_METHOD
-                );
-            } else {
-                // duplicate target
-                writer.dup();
-                pushCastToType(writer, declaringTypeObject);
-            }
-
-            boolean hasArgs = !argumentTypes.isEmpty();
-            if (hasArgs) {
-                if (reflectionRequired) {
+                writer.invokeVirtual(ExecutableMethodsDefinitionWriter.SUPER_TYPE, GET_ACCESSIBLE_TARGET_METHOD);
+                if (hasArgs) {
                     writer.loadArg(2);
                 } else {
+                    writer.getStatic(Type.getType(ArrayUtils.class), "EMPTY_OBJECT_ARRAY", Type.getType(Object[].class));
+                }
+                writer.invokeStatic(TYPE_REFLECTION_UTILS, METHOD_INVOKE_METHOD);
+            } else {
+                pushCastToType(writer, declaringTypeObject);
+                if (hasArgs) {
                     int argCount = argumentTypes.size();
                     Iterator<ParameterElement> argIterator = argumentTypes.iterator();
                     for (int i = 0; i < argCount; i++) {
@@ -532,25 +528,15 @@ public class DispatchWriter extends AbstractClassFileWriter implements Opcodes {
                         pushCastToType(writer, argIterator.next());
                     }
                 }
-            } else if (reflectionRequired) {
-                writer.getStatic(Type.getType(ArrayUtils.class), "EMPTY_OBJECT_ARRAY", Type.getType(Object[].class));
-            }
-
-            if (reflectionRequired) {
-                writer.invokeStatic(
-                        TYPE_REFLECTION_UTILS,
-                        METHOD_INVOKE_METHOD
-                );
-            } else {
                 String methodDescriptor = getMethodDescriptor(returnType, argumentTypes);
                 writer.visitMethodInsn(isInterface ? INVOKEINTERFACE : INVOKEVIRTUAL,
-                                       declaringTypeObject.getInternalName(), methodName,
-                                       methodDescriptor, isInterface);
+                    declaringTypeObject.getInternalName(), methodName,
+                    methodDescriptor, isInterface);
             }
 
             if (returnTypeObject.equals(Type.VOID_TYPE)) {
                 writer.visitInsn(ACONST_NULL);
-            } else {
+            } else if (!reflectionRequired) {
                 pushBoxPrimitiveIfNecessary(returnType, writer);
             }
         }

